File: howl-api/src/main/java/io/quartic/howl/api/HowlService.java
package io.quartic.howl.api;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Optional;
import java.util.function.Consumer;

public interface HowlService {
    void uploadFile(String contentType, String namespace, String fileName, Consumer<OutputStream> upload) throws IOException;
    HowlStorageId uploadFile(String contentType, String namespace, Consumer<OutputStream> upload) throws IOException;
    Optional<InputStream> downloadFile(String namespace, String fileName) throws IOException;
}

File: howl-api/src/main/kotlin/io/quartic/howl/api/HowlService.kt
package io.quartic.howl.api

import java.io.IOException
import java.io.InputStream
import java.io.OutputStream

interface HowlService {
    @Throws(IOException::class)
    fun uploadFile(
        targetNamespace: String,
        key: String,
        contentType: String,
        upload: (OutputStream) -> Unit
    )

    @Throws(IOException::class)
    fun uploadFile(
        targetNamespace: String,
        identityNamespace: String,
        key: String,
        contentType: String,
        upload: (OutputStream) -> Unit
    )

    @Throws(IOException::class)
    fun uploadAnonymousFile(
        targetNamespace: String,
        contentType: String,
        upload: (OutputStream) -> Unit
    ): HowlStorageId

    @Throws(IOException::class)
    fun uploadAnonymousFile(
        targetNamespace: String,
        identityNamespace: String,
        contentType: String,
        upload: (OutputStream) -> Unit
    ): HowlStorageId

    @Throws(IOException::class)
    fun downloadFile(targetNamespace: String, key: String): InputStream?

    @Throws(IOException::class)
    fun downloadFile(targetNamespace: String, identityNamespace: String, key: String): InputStream?
}
